src/ui/SortDropdown/SortDropdown.ts

```diff
@@ -0,0 +1,159 @@
+import { Sort } from '../Sort/Sort';
+import { SortCriteria } from '../Sort/SortCriteria';
+import { Dropdown } from '../FormWidgets/Dropdown';
+import { l } from '../../strings/Strings';
+
+export interface ISortDropdownItem {
+  criteria: SortCriteria;
+  caption: string;
+}
+
+export interface ISortDropdownOptions {}
+
+/**
+ * The `SortDropdown` component renders a dropdown that the end user can interact with to select the criterion to use when sorting query results.
+ */
+export class SortDropdown extends Component {
+  static ID = 'SortDropdown';
+
+  static doExport = () => {
+    exportGlobally({
+      SortDropdown: SortDropdown
+    });
+  };
+
+  /**
+   * Options for the component
+   * @componentOptions
+   */
+  static options: ISortDropdownOptions = {};
+
+  private dropdown: Dropdown;
+  private sortItems: ISortDropdownItem[] = [];
+
+  /**
+   * Creates a new `SortDropdown` component instance.
+   * @param element The HTMLElement on which to instantiate the component.
+   * @param options The options for this component instance.
+   * @param bindings The bindings that the component requires to function normally. If not set, these will be
+   * automatically resolved (with a slower execution time).
+   */
+  constructor(public element: HTMLElement, public options?: ISortDropdownOptions, bindings?: IComponentBindings) {
+    super(element, SortDropdown.ID, bindings);
+
+    this.options = ComponentOptions.initComponentOptions(element, SortDropdown, options);
+
+    this.bind.oneRootElement(InitializationEvents.afterInitialization, () => this.handleAfterInitialization());
+    this.bind.onQueryState(MODEL_EVENTS.CHANGE_ONE, QUERY_STATE_ATTRIBUTES.SORT, (args: IAttributeChangedEventArg) =>
+      this.handleQueryStateChanged(args)
+    );
+    this.bind.onRootElement(QueryEvents.querySuccess, (args: IQuerySuccessEventArgs) => this.handleQuerySuccess(args));
+    this.bind.onRootElement(QueryEvents.queryError, (args: IQueryErrorEventArgs) => this.handleQueryError(args));
+  }
+
+  private handleAfterInitialization() {
+    this.buildDropdown();
+  }
+
+  private buildDropdown() {
+    this.sortItems = this.getSortItems();
+    this.dropdown = new Dropdown(
+      () => this.handleChange(),
+      this.getValuesForDropdown(),
+      value => this.getCaptionForValue(value),
+      l('SortBy')
+    );
+    this.element.appendChild(this.dropdown.getElement());
+    this.update();
+  }
+
+  private getSortItems() {
+    const sortItems: ISortDropdownItem[] = [];
+    const sortElements = $$(this.element).findAll('.CoveoSort');
+
+    each(sortElements, sortEl => {
+      const sortInstance = <Sort>Component.get(sortEl, Sort);
+
+      if (sortInstance.options.sortCriteria.length > 1) {
+        this.logger.warn(
+          `Each Sort component inside a SortDropdown should have only one sort criteria. Not using ${
+            sortInstance.options.caption
+          } in the SortDropdown.`
```

Comment:
While I was initially for composition, I think it would be better to not go down this path due to this warning as well as IEditor drag-and-drop considerations down the line.

Thinking of the IEditor user experience, it seems cumbersome to require a person to first drop a SortDropdown component, and then create/drop individual Sort components as suggested by composition.

Another issue is we would need to put in additional effort to allow accessing/editing the Sort components inside the SortDropdown since they would not be visible (we would need an analogue to the Result Template editor). Changing the styling would to display everything in Edit mode would not be good since it breaks the WYSIWYG principle of the project.

Add to that needing to restrict adding a second sort order when editing a Sort component inside a SortDropdown, as well as preventing someone from dropping a Sort Component with two sort orders configured.


Using a table if the setting remains of JSON type (or possibly using two string arrays) seems like a simpler solution versus the composition approach.

Reply:
I kind of get your point here Sami, but if we think about the fact that the interface editor is only "a part" of all users that will end up configuring this component (for example, in commerce, no clients are using interface editor right now...), we are effectively making the configuration for our users using "code only" worse or more cumbersome.

I think we should strive to make the "base" layer (aka: the code config) as natural/simple as possible, and then see how the interface editor could adapt itself, and not the other way around.

I don't see how using JSON option would "really" make this simpler for the interface editor also ? 

Meaning, if we decide that the way to configure that in the interface editor using some kind of table view, then "how" the table is populated and edited and the underlying "data" in the HTML that represents that is just implementing a different kind of plumbing/parsing ? And the "view" we end up building for this component would not change even if we modified how the component options is built in the JSUI.

And the same restriction/warning that is shown here would still apply, no ?

Reply:
Perfect, I was imagining users editing dragging individual sort components into/out of the SortDropdown. If instead we use a table and generate the correct mark-up, this would allow keeping both the code and end-user configuration simple.